File: epi/Python/bst_12.py
# Convert a BST to a sorted doubly linked list
# Design an algorithm that takes as input a BST B an returns a sorted doubly linked list on the
# same elements. Your algorithm should not allocate any new nodes. The original BST does not have
# to be preserved; use its nodes as the nodes of the resulting list
def bst_to_linked_list(root):
    l = Node(None)
    first = bst_to_linked_list_helper(root)[0]
    l.right = first
    return l
def bst_to_linked_list_helper(root):
    if root.left == None and root.right == None:
        return (root, root)
    left = root
    right = root
    if root.left:
        res1 = bst_to_linked_list_helper(root.left)
        root.left = res1[1]
        res1[1].right = root
        left = res1[0]
    if root.right:
        res2 = bst_to_linked_list_helper(root.right)
        root.right = res2[0]
        res2[0].left = root
        right = res2[1]
    return (left, right)

class Node():
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

    def __str__(self):
        curr = self
        res = str(self.data)
        while curr.right:
            curr = curr.right
            res += " -> " + str(curr.data)
        return res

if __name__ == "__main__":
    a = Node(7); b = Node(3); c = Node(11); d = Node(2); e = Node(5)
    a.left = b; a.right = c; b.left = d; b.right = e
    l = bst_to_linked_list(a)
    print l.right

File: epi/Ruby/bst_12.rb
=begin
Convert a BST to a sorted doubly linked list
=end
require "libraries"
def convert(root)
  head = TNode.new(nil)
  head.right = convert_helper(root)[:left]
  return head
end

def convert_helper(root)
  if root.nil?
    return {left: nil, right: nil}
  end
  if root.left.nil? && root.right.nil?
    return {left: root, right: root}
  end
  left = convert_helper(root.left)
  right = convert_helper(root.right)
  root.left = left[:right]
  root.right = right[:left]
  left[:right].right = root if left[:right]
  right[:left].left = root if right[:left]
  return {left: left[:left] ? left[:left] : root,
    right: right[:right] ? right[:right] : root}
end

def right_ward(head)
  res = []
  return res if head.nil?
  head = head.right
  while head
    res << head.data
    head = head.right
  end
  res
end

root = TNode.new(7)
root.left = TNode.new(3)
root.right = TNode.new(11)
root.left.left = TNode.new(2)
root.left.right = TNode.new(5)
head = convert(root)
puts right_ward(head).to_s
